kostyor/upgrades/driver.py

```diff
@@ -0,0 +1,85 @@
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class UpgradeDriver():
+
+    @abc.abstractproperty
+    def supports_upgrade_rollback(self):
+        """Property used to indicate if an implementation supports rolling back
+        an upgrade
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_host_upgrade_hook(self, host):
+        """Called by the decision engine before a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_service_upgrade_hook(self, service):
+        """Called by the decision engine before a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_host_upgrade_hook(self, host):
+        """Called by the decision engine after a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_service_upgrade_hook(self, service):
+        """Called by the decision engine after a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def cancel_upgrade(self):
```

Comment:
Hm, it's a good question. Considering that we may have multiple clusters with different drivers it doesn't seem like a singleton is an option here. Besides, we usually have multiple API processes so each instance will have its own singleton.

p.s: or by singleton you mean no state and just pass everything to class methods?